File: templates/Enterprise-Template/src/csharp/EnterpriseBotTemplate/Bot Framework/Dialogs/Onboarding/OnboardingDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using $safeprojectname$.Dialogs.Shared;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace $safeprojectname$.Dialogs.Onboarding
{
    public class OnboardingDialog : EnterpriseDialog
    {
        private static OnboardingResponses _responder = new OnboardingResponses();
        private IStatePropertyAccessor<OnboardingState> _accessor;
        private OnboardingState _state;

        public OnboardingDialog(BotServices botServices, IStatePropertyAccessor<OnboardingState> accessor)
            : base(botServices, nameof(OnboardingDialog))
        {
            _accessor = accessor;
            InitialDialogId = nameof(OnboardingDialog);

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                AskForEmail,
                AskForLocation,
                FinishOnboardingDialog,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, onboarding));
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
            AddDialog(new TextPrompt(DialogIds.EmailPrompt));
            AddDialog(new TextPrompt(DialogIds.LocationPrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());

            if (!string.IsNullOrEmpty(_state.Name))
            {
                return await sc.NextAsync(_state.Name);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.NamePrompt, new PromptOptions()
                {
                    Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.NamePrompt),
                });
            }
        }

        public async Task<DialogTurnResult> AskForEmail(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            var name = _state.Name = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveNameMessage, new { name });

            return await sc.PromptAsync(DialogIds.EmailPrompt, new PromptOptions()
            {
                Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.EmailPrompt),
            });
        }

        public async Task<DialogTurnResult> AskForLocation(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            var email = _state.Email = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveEmailMessage, new { email });

            if (!string.IsNullOrEmpty(_state.Location))
            {
                return await sc.NextAsync(_state.Location);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.LocationPrompt, new PromptOptions()
                {
                    Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.LocationPrompt),
                });
            }
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context);
            _state.Location = (string)sc.Result;

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveLocationMessage, new { _state.Name, _state.Location });
            return await sc.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
            public const string EmailPrompt = "emailPrompt";
            public const string LocationPrompt = "locationPrompt";
        }
    }
}

File: templates/Enterprise-Template/src/typescript/src/dialogs/onboarding/onboardingDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { StatePropertyAccessor, TurnContext } from "botbuilder";
import { DialogTurnResult, TextPrompt, WaterfallDialog, WaterfallStepContext } from "botbuilder-dialogs";
import { BotServices } from "../../botServices";
import { EnterpriseDialog } from "../shared/enterpriseDialog";
import { OnboardingResponses } from "./onboardingResponses";
import { OnboardingState } from "./onboardingState";

export class OnboardingDialog extends EnterpriseDialog {

    // Fields
    private readonly _responder: OnboardingResponses;
    private readonly _accessor: StatePropertyAccessor<OnboardingState>;
    private _state!: OnboardingState;

    constructor(botServices: BotServices, accessor: StatePropertyAccessor<OnboardingState>) {
        super(botServices, OnboardingDialog.name);

        this._accessor = accessor;
        this.initialDialogId = OnboardingDialog.name;
        this._responder = new OnboardingResponses();

        const onboarding = [
            this.askForName.bind(this),
            this.askForEmail.bind(this),
            this.askForLocation.bind(this),
            this.finishOnboardingDialog.bind(this),
        ];

        this.addDialog(new WaterfallDialog<OnboardingState>(this.initialDialogId, onboarding));
        this.addDialog(new TextPrompt(DialogIds.NamePrompt));
        this.addDialog(new TextPrompt(DialogIds.EmailPrompt));
        this.addDialog(new TextPrompt(DialogIds.LocationPrompt));
    }

    public async askForName(sc: WaterfallStepContext<OnboardingState>): Promise<DialogTurnResult> {
        this._state = await this.getStateFromAccessor(sc.context);

        if (this._state.name) {
            return await sc.next(this._state.name);
        } else {
            return await sc.prompt(DialogIds.NamePrompt, {
                prompt: await this._responder.renderTemplate(sc.context, OnboardingResponses.ResponseIds.NamePrompt, sc.context.activity.locale as string),
            });
        }
    }

    public async askForEmail(sc: WaterfallStepContext<OnboardingState>): Promise<DialogTurnResult> {
        this._state = await this.getStateFromAccessor(sc.context);
        this._state.name = sc.result;

        await this._responder.replyWith(sc.context, OnboardingResponses.ResponseIds.HaveNameMessage, { name: this._state.name });

        return await sc.prompt(DialogIds.EmailPrompt, {
            prompt: await this._responder.renderTemplate(sc.context, OnboardingResponses.ResponseIds.EmailPrompt, sc.context.activity.locale as string),
        });
    }

    public async askForLocation(sc: WaterfallStepContext<OnboardingState>): Promise<DialogTurnResult> {
        this._state = await this.getStateFromAccessor(sc.context);
        this._state.email = sc.result;

        await this._responder.replyWith(sc.context, OnboardingResponses.ResponseIds.HaveEmailMessage, { email: this._state.email });

        return await sc.prompt(DialogIds.LocationPrompt, {
            prompt: await this._responder.renderTemplate(sc.context, OnboardingResponses.ResponseIds.LocationPrompt, sc.context.activity.locale as string),
        });     
    }

    public async finishOnboardingDialog(sc: WaterfallStepContext<OnboardingState>): Promise<DialogTurnResult> {
        this._state = await this.getStateFromAccessor(sc.context);
        this._state.location = sc.result as string;

        await this._responder.replyWith(sc.context, OnboardingResponses.ResponseIds.HaveLocationMessage, { name: this._state.name, location: this._state.location });

        return await sc.endDialog();
    }

    private async getStateFromAccessor(context: TurnContext): Promise<OnboardingState>  {
        const state: OnboardingState | undefined = await this._accessor.get(context);
        if (!state) {
            const newState: OnboardingState = {
                email: "",
                location: "",
                name: "",
            };
            await this._accessor.set(context, newState);
            return newState;
        }
        return state;
    }
}

class DialogIds {
    public static NamePrompt: string = "namePrompt";
    public static EmailPrompt: string = "emailPrompt";
    public static LocationPrompt: string =  "locationPrompt";
}